sdk/core/AzureCore/Source/Pipeline/Transport/QueryParameters.swift

```diff
@@ -0,0 +1,91 @@
+// --------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation. All rights reserved.
+//
+// The MIT License (MIT)
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the ""Software""), to
+// deal in the Software without restriction, including without limitation the
+// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+// sell copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//
+// --------------------------------------------------------------------------
+
+import Foundation
+
+public protocol QueryStringConvertible {
+    var queryValue: String { get }
+}
+
+public typealias QueryParameter = (key: String, value: String)
+
+public struct QueryParameters {
+    internal var parameters = [QueryParameter]()
+
+    public init(_ params: [(key: String, value: QueryStringConvertible?)]? = nil) {
+        for param in params ?? [] {
+            add(key: param.key, value: param.value)
+        }
+    }
+
+    public mutating func add(key: String, value: QueryStringConvertible?) {
+        // skip values that evaluate to nil
+        guard let val = value else { return }
+        parameters.append((key: key, value: val.queryValue))
+    }
+}
+
+extension Int: QueryStringConvertible {
```

Comment:
I initially tried an approach taking an array of `[AnyObject]` and doing a switch on the type. However, this failed because the value 1 of type `Any` successfully casts to a `Bool` and an `Int`, making it impossible to meaningfully distinguish between the two. It essentially boils down to whichever case statement is listed first.

I opted for `QueryStringConvertible` over `CustomStringConvertible` since someone might want to use `CustomStringConvertible` for other, more general purposes and not have that hijacked by us.

The other implementation I pursued was having many overloads of the `add` method, but this resulted in a lot of methods when you dot on the struct. Thus, having a single method that takes a protocol type seemed like the cleanest implementation.

The purpose of this is essentially to relocate the complexity of generating the right string from the generated code (via Autorest.Swift) into AzureCore, which is much much easier to do because of the limits of the Swift Stencil framework.

While these extensions seem awfully special case, they are special cased based on what the testserver expects us to send on the wire in these instances.